### PR TITLE
Improve modeling of test fixtures

### PR DESCRIPTION
**Pull request by oehme**
Previously they were modeled as just a file collection,
which confused the IDEA importer. They now use the more
canonical way of defining an additional artifact.

In a next step they should be reimplemented using variant
aware dependency resolution, so that we don't need a special
DSL to declare these dependencies.

cc @wolfs as promised yesterday
cc @eskatos because I'm still a Kotlin noob.

### CODE REVIEW

File: buildSrc/subprojects/integration-testing/src/main/kotlin/org/gradle/gradlebuild/test/fixtures/TestFixturesPlugin.kt [hunk 1]
```diff
@@ -60,46 +55,72 @@ open class TestFixturesPlugin : Plugin<Project> {
...
+    /**
+     * This mimics what the java-library plugin does, but creating a library of test fixtures instead.
+     */
     private
     fun Project.configureAsProducer() {
+        val testFixtures by java.sourceSets.creating
 
-        configurations {
-            create("outputDirs")
+        java.sourceSets.named<SourceSet>("test") {
```
Comment (oehme) [inline buildSrc/subprojects/integration-testing/src/main/kotlin/org/gradle/gradlebuild/test/fixtures/TestFixturesPlugin.kt]: Is there really no pretty way? It feels inconsistent not to use `by getting` here, but then Idea complains about an unused variable.

File: buildSrc/subprojects/integration-testing/src/main/kotlin/org/gradle/gradlebuild/test/fixtures/TestFixturesPlugin.kt [hunk 2]
```diff
@@ -60,46 +55,72 @@ open class TestFixturesPlugin : Plugin<Project> {
...
+            compileClasspath += testFixtures.output
+            runtimeClasspath += testFixtures.output
+        }
 
-            create("testFixturesCompile") { extendsFrom(configurations["compile"]) }
-            create("testFixturesImplementation") { extendsFrom(configurations["implementation"]) }
-            create("testFixturesRuntime") { extendsFrom(configurations["runtime"]) }
+        val testFixturesJar by tasks.creating(Jar::class) {
```
Comment (oehme) [inline buildSrc/subprojects/integration-testing/src/main/kotlin/org/gradle/gradlebuild/test/fixtures/TestFixturesPlugin.kt]: Are we going to get rid of the :: syntax?

Comment (eskatos) [reply buildSrc/subprojects/integration-testing/src/main/kotlin/org/gradle/gradlebuild/test/fixtures/TestFixturesPlugin.kt]: The `::class` syntax is the canonical way to pass types around in Kotlin, so we won't get rid of it entirely 😄 

See https://github.com/gradle/kotlin-dsl/issues/605

File: buildSrc/subprojects/integration-testing/src/main/kotlin/org/gradle/gradlebuild/test/fixtures/TestFixturesPlugin.kt (hunk 1 above)
Comment (eskatos) [reply buildSrc/subprojects/integration-testing/src/main/kotlin/org/gradle/gradlebuild/test/fixtures/TestFixturesPlugin.kt]: The `SourceSetContainer` is now available as an extension on `Project`. Accessing it that way would shorten this to `sourceSets.named<SourceSet>("test") {`.

The container isn't polymorphic, it should be possible to say `sourceSets.named("test") {`. But it's not, the receiver is then typed `Any`. This is a gap in the kotlin-dsl sugar, thanks! Opened https://github.com/gradle/kotlin-dsl/issues/1104

**Comment by wolfs** [comment]
❌  Compilation doesn't seem to work anymore: https://e.grdev.net/s/3cetc3ost6uze/console-log?task=:buildInit:compileIntegTestGroovy

File: buildSrc/subprojects/integration-testing/src/main/kotlin/org/gradle/gradlebuild/test/fixtures/TestFixturesPlugin.kt [hunk 3]
```diff
@@ -60,46 +55,72 @@ open class TestFixturesPlugin : Plugin<Project> {
...
+     * This mimics what the java-library plugin does, but creating a library of test fixtures instead.
+     */
     private
     fun Project.configureAsProducer() {
+        val testFixtures by java.sourceSets.creating
 
-        configurations {
-            create("outputDirs")
+        java.sourceSets.named<SourceSet>("test") {
+            compileClasspath += testFixtures.output
```
Comment (wolfs) [inline buildSrc/subprojects/integration-testing/src/main/kotlin/org/gradle/gradlebuild/test/fixtures/TestFixturesPlugin.kt]: Would there be any advantages in adding the `jar` as a dependency instead of the output of the source set?

File: buildSrc/subprojects/integration-testing/src/main/kotlin/org/gradle/gradlebuild/test/fixtures/TestFixturesPlugin.kt [hunk 4]
```diff
@@ -60,46 +55,72 @@ open class TestFixturesPlugin : Plugin<Project> {
...
+            from(testFixtures.output)
+            classifier = "test-fixtures"
+        }
 
-            // Expose configurations that include the test fixture classes for clients to use
-            create("testFixturesUsageCompile") {
-                extendsFrom(configurations["testFixturesCompile"], configurations["outputDirs"])
+        configurations {
+            val testFixturesCompile by getting {
+                extendsFrom(configurations["compile"])
```
Comment (wolfs) [inline buildSrc/subprojects/integration-testing/src/main/kotlin/org/gradle/gradlebuild/test/fixtures/TestFixturesPlugin.kt]: ❌ Shouldn't we use the constants on `JavaPlugin` instead of plain strings?

File: buildSrc/subprojects/integration-testing/src/main/kotlin/org/gradle/gradlebuild/test/fixtures/TestFixturesPlugin.kt [hunk 5]
```diff
@@ -109,32 +130,18 @@ open class TestFixturesPlugin : Plugin<Project> {
                 }
             }
         }
-
-        plugins.withType<EclipsePlugin> {
```
Comment (wolfs) [inline buildSrc/subprojects/integration-testing/src/main/kotlin/org/gradle/gradlebuild/test/fixtures/TestFixturesPlugin.kt]: So eclipse should work out of the box now?

**Comment by wolfs** [comment]
By these changes, the projects using the test fixtures are still consuming a configuration. The only difference is that the configuration now contains a `jar` instead of the output directories.

I was somewhat expecting that they would consume a variant of the project containing the test fixtures.

I don't really understand why consuming the jar is better for IDE support. Since the Gradle output directories are not there any more?

File: buildSrc/subprojects/integration-testing/src/main/kotlin/org/gradle/gradlebuild/test/fixtures/TestFixturesPlugin.kt (hunk 3 above)
Comment (oehme) [reply buildSrc/subprojects/integration-testing/src/main/kotlin/org/gradle/gradlebuild/test/fixtures/TestFixturesPlugin.kt]: No, but I'm gonna rework this anyway :)

File: buildSrc/subprojects/integration-testing/src/main/kotlin/org/gradle/gradlebuild/test/fixtures/TestFixturesPlugin.kt (hunk 4 above)
Comment (oehme) [reply buildSrc/subprojects/integration-testing/src/main/kotlin/org/gradle/gradlebuild/test/fixtures/TestFixturesPlugin.kt]: yep

File: buildSrc/subprojects/integration-testing/src/main/kotlin/org/gradle/gradlebuild/test/fixtures/TestFixturesPlugin.kt (hunk 5 above)
Comment (oehme) [reply buildSrc/subprojects/integration-testing/src/main/kotlin/org/gradle/gradlebuild/test/fixtures/TestFixturesPlugin.kt]: yes

**Comment by oehme** [comment]
> I was somewhat expecting that they would consume a variant of the project containing the test fixtures.

That's what I'm doing next.

> I don't really understand why consuming the jar is better for IDE support. Since the Gradle output directories are not there any more?

The main problem was making the project depend on its `testFixtures.output`, which won't be recognized properly by any IDE and led to the problems you were seeing.

**Comment by oehme** [comment]
@wolfs I'm splitting this work into two parts: This is the first part which should make the IDEA importer happy. It's also the "old" idiomatic way of defining additional artifacts.

In another PR I'll reimplement them using variant aware dependency resolution. This turned out to be a bigger problem as we have some nasty cycles. E.g. `core` depends on `messaging`, but `messaging`  depends on `core`'s test fixtures. This clearly means that `core` contains fixtures that should live in `base-services` instead.

File: buildSrc/subprojects/integration-testing/src/main/kotlin/org/gradle/gradlebuild/test/fixtures/TestFixturesPlugin.kt [hunk 6]
```diff
@@ -110,32 +137,36 @@ open class TestFixturesPlugin : Plugin<Project> {
...
     fun Project.configureAsConsumer() = afterEvaluate {
-
         the<TestFixturesExtension>().origins.forEach { (projectPath, sourceSetName) ->
-
             val compileConfig = if (sourceSetName == "main") "compile" else "${sourceSetName}Compile"
             val runtimeConfig = if (sourceSetName == "main") "runtime" else "${sourceSetName}Runtime"
 
             dependencies {
-                compileConfig(project(path = projectPath, configuration = "testFixturesUsageCompile"))
+                compileConfig(project(path = projectPath, configuration = "testFixturesApiElements"))
```
Comment (wolfs) [inline buildSrc/subprojects/integration-testing/src/main/kotlin/org/gradle/gradlebuild/test/fixtures/TestFixturesPlugin.kt]: Shall we extract those two configuration names as a constant?
